nexus/src/db/subnet_allocation.rs

```diff
@@ -219,9 +224,278 @@ impl QueryFragment<Pg> for AllocateIpQueryValues {
     }
 }
 
+/// Generate a CTE that can be used to insert a VPC Subnet, only if the IP
+/// address ranges of that subnet don't overlap with existing Subnets in the
+/// same VPC.
+///
+/// In particular, this generates a CTE like so:
+///
+/// ```sql
+/// WITH candidate(
+///     id,
+///     name,
+///     description,
+///     time_created,
+///     time_modified,
+///     time_deleted,
+///     vpc_id,
+///     ipv4_block,
+///     ipv6_block
+/// ) AS (VALUES (
+///     <id>,
+///     <name>,
+///     <description>,
+///     <time_created>,
+///     <time_modified>,
+///     NULL::TIMESTAMPTZ,
+///     <vpc_id>,
+///     <ipv4_block>,
+///     <ipv6_block>
+/// ))
+/// SELECT *
+/// FROM candidate
+/// WHERE NOT EXISTS (
+///     SELECT ipv4_block, ipv6_block
+///     FROM vpc_subnet
+///     WHERE
+///         vpc_id = <vpc_id> AND
+///         time_deleted IS NULL AND
+///         (
+///             (ipv4_block && candidate.ipv4_block) OR
```

Comment:
So, the asymptotics first. Yes, I believe this will be linear in the number of VPC Subnets in a VPC. I also agree that should be fine for a long time, until people really are creating thousands of subnets. Hopefully 🤞 

As for the operator. That's not a bitwise comparison, that's actually a [network address function](https://www.postgresql.org/docs/14/functions-net.html) that I believe came from Postgres. Specifically, it tests for "symmetric containment," as in, does either subnet contain or equal the other. Here is [the actual function](https://www.postgresql.org/docs/14/functions-net.html) that CRDB uses. I played with it in the shell enough to convince myself it's doing the right thing, but I'd be happy to add some experimentation notes or examples to this discussion.

Reply:
So, here's your exact example in the CRDB SQL shell:

```
root@127.0.0.1:32221/omicron> SELECT '10.0.0.0/9'::INET && '10.128.0.0/9'::INET;
  ?column?
------------
   false
(1 row)

Time: 0ms total (execution 0ms / network 0ms)
```

Some more examples, which I think show the right behavior:

```
root@127.0.0.1:32221/omicron> SELECT '10.0.0.0/9'::INET && '10.128.0.0/10'::INET;
  ?column?
------------
   false
(1 row)

Time: 0ms total (execution 0ms / network 0ms)

root@127.0.0.1:32221/omicron> SELECT '10.0.0.0/9'::INET && '10.1.0.0/9'::INET;
  ?column?
------------
    true
(1 row)

Time: 0ms total (execution 0ms / network 0ms)

root@127.0.0.1:32221/omicron> SELECT '10.1.0.0/9'::INET && '10.0.0.0/9'::INET;
  ?column?
------------
    true
(1 row)

Time: 0ms total (execution 0ms / network 0ms)

root@127.0.0.1:32221/omicron> SELECT '10.0.0.0/9'::INET && '10.0.1.0/9'::INET;
  ?column?
------------
    true
(1 row)

Time: 0ms total (execution 0ms / network 0ms)

root@127.0.0.1:32221/omicron> SELECT '1.0.0.0/9'::INET && '10.0.0.0/9'::INET;
  ?column?
------------
   false
(1 row)

Time: 0ms total (execution 0ms / network 0ms)
```

Reply:
Will comment, or maybe use the explicit function `inet_contains_or_equals`.

Reply:
I've used the `inet_contains_or_equals` function explicitly.